File: CodeForces/goodbye_2020_F.py
from collections import defaultdict

class UF:
    def __init__(self, N):
        self.par = list(range(N))
        self.sz = [1] * N

    def find(self, x):
        if self.par[x] != x:
            self.par[x] = self.find(self.par[x])
        return self.par[x]

    def union(self, x, y):
        xr, yr = self.find(x), self.find(y)
        if xr == yr:
            return False
        if self.sz[xr] < self.sz[yr]:
            xr, yr = yr, xr
        self.par[yr] = xr
        self.sz[xr] += self.sz[yr]
        return True

    def size(self, x):
        return self.sz[self.find(x)]

    def connected(self, x, y):
        return self.find(x) == self.find(y)

n,m = list(map(int, input().split(' ')))
cur_count = 1
cur_lst = []
uf = UF(m+2)
for i in range(n):
    bit_lst = list(map(int, input().split(' ')))
    bit1 = bit_lst[1]
    bit2 = m+1
    if bit_lst[0] == 2:
        bit2 = bit_lst[2]
    if uf.union(bit1, bit2):
        cur_lst.append(str(i+1))
        cur_count *= 2 
        cur_count %= 10**9 + 7

print(cur_count, len(cur_lst))
print(" ".join(cur_lst))

File: CodeForces/goodbye_2020/goodbye_2020_F.cc
#include <bits/stdc++.h>

using namespace std;

const int N = 5e5 + 7;
const int MX = 1e9 + 7;

int uf[N];
int sz[N];

int find(int x){
    if (uf[x] != x){
       uf[x] = find(uf[x]);
    }
    return uf[x];
}

bool _union(int x, int y){
    int xr = find(x);
    int yr = find(y);

    if (xr == yr){
        return false;
    }
    if (sz[xr] < sz[yr]){
        xr, yr = yr, xr;
    }
    uf[yr] = xr;
    sz[xr] += sz[yr];
    return true;
}

int main(){
    int n,m;
    int ans = 1;
    scanf("%d %d", &n, &m);
    for (int i = 1; i <= m+1; ++i){
        uf[i] = i;
        sz[i] = 1;
    }
    vector<int> cur_lst;
    for (int i = 1; i <= n; ++i){
        int bit_lst_size, bit1, bit2;
        bit1 = bit2 = m+1;
        scanf("%d", &bit_lst_size);
        scanf("%d", &bit1);
        if (bit_lst_size == 2){
            scanf("%d", &bit2);
        }
        if (_union(bit1, bit2)){
            cur_lst.push_back(i);
            ans *= 2;
            ans %= MX;
        }
    }
    printf("%d %d\n", ans, cur_lst.size());
    for (auto v: cur_lst){
        printf("%d ", v);
    }
    return 0;
}